gssapi.go

```diff
@@ -0,0 +1,130 @@
+package ldap
+
+import (
+	"fmt"
+
+	"github.com/jcmturner/gokrb5/v8/client"
+	"github.com/jcmturner/gokrb5/v8/crypto"
+	"github.com/jcmturner/gokrb5/v8/iana/keyusage"
+	"github.com/jcmturner/gokrb5/v8/messages"
+	"github.com/jcmturner/gokrb5/v8/types"
+
+	"github.com/jcmturner/gokrb5/v8/gssapi"
+	"github.com/jcmturner/gokrb5/v8/spnego"
+)
+
+type GSSAPIState struct {
+	AuthZID string
+	token   spnego.KRB5Token
+	ekey    types.EncryptionKey
+	Subkey  types.EncryptionKey
+	init    bool
+	asrep   bool
+}
+
+func InitContext(client *client.Client, principal, AuthZID string) (*GSSAPIState, error) {
+	tkt, ekey, err := client.GetServiceTicket(principal)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := spnego.NewKRB5TokenAPREQ(client, tkt, ekey, []int{gssapi.ContextFlagInteg, gssapi.ContextFlagConf, gssapi.ContextFlagMutual}, []int{})
+	if err != nil {
+		return nil, err
+	}
+
+	state := &GSSAPIState{
+		AuthZID: AuthZID,
+		ekey:    ekey,
+		token:   token,
+		init:    false,
+		asrep:   false,
+	}
+
+	return state, nil
+}
+
+func (state *GSSAPIState) GSSAPIStep(input []byte) ([]byte, error) {
+	if !state.init {
+		state.init = true
+		return state.token.Marshal()
+	}
+
+	if !state.asrep {
+		err := state.token.Unmarshal(input)
+		if err != nil {
+			return nil, err
+		}
+
+		if state.token.IsAPRep() {
+			state.asrep = true
+
+			encpart, err := crypto.DecryptEncPart(state.token.APRep.EncPart, state.ekey, keyusage.AP_REP_ENCPART)
+			if err != nil {
+				return nil, err
+			}
+
+			part := &messages.EncAPRepPart{}
+			err = part.Unmarshal(encpart)
+			if err != nil {
+				return nil, err
+			}
+
+			state.Subkey = part.Subkey
+		}
+
+		if state.token.IsKRBError() {
+			return nil, state.token.KRBError
+		}
+
+		return make([]byte, 0), nil
+	}
+
+	token := &gssapi.WrapToken{}
+	err := token.Unmarshal(input, true)
```

Comment:
testing against AD, this only works if I hacked the Unmarshal() call, which could be done here, after the fact, instead of within Unmarshal() itself, if needed. otherwise the call to token.Verify() fails w/ a checksum mismatch.

the hack: Unmarshal() parses `header, payload, checksum` in that order. to get this working with AD, I need to change the parse order: `header, checksum, payload`.